adb/common.py

```diff
@@ -292,22 +294,34 @@ def __init__(self, serial):
     else:
       host = serial
       port = 5555
+    self._timeout_ms = timeout_ms
     self._serial_number = '%s:%s' % (host, port)
-
-    self._connection = socket.create_connection((host, port))
+    self._timeout_ms = float(timeout_ms)
+    timeout = float(timeout_ms)/1000.0 if timeout_ms else None 
+    self._connection = socket.create_connection((host,port),timeout=timeout)
+    if timeout:
+        self._connection.setblocking(0)
 
   @property
   def serial_number(self):
     return self._serial_number
 
   def BulkWrite(self, data, timeout=None):
+      if self._timeout_ms:
+          return self._connection.send(data)
```

Comment:
what's the difference here? What if we did `send` in the other case too?

Reply:
My initial concern was in that self._connection.sendall(data) requires a blocking connection as sendall is supposed to send all data . Connection is set to non blocking (self._connection.setblocking(0)) to allow for timeout on the connection. From these points it seems self._connection.send() is the way to go, however the docs on "send()" says: "Applications are responsible for checking that all data has been sent": From what I gather (excuse my ignorance) it seems such a check is not done (adb_protocol.Send(usb, timeout)).